onnxruntime/core/session/inference_session.cc

```diff
@@ -415,96 +415,65 @@ class InferenceSession::Impl {
       if (feed_names_entry == end_names) {
-        continue;
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                               "Missing required input: ", arg_name);
       }
 
       auto idx = feed_names_entry - begin_names;
+      required_feed_ids.insert(idx);
       auto& input_ml_value = feeds.at(idx);
-      auto input_type = input_ml_value.Type();
       auto expected_type = utils::GetMLDataType(*arg);
 
-      if (!input_ml_value.IsTensor()) {
-        auto retval = CheckTypes(input_type, expected_type);
-        if (!retval.IsOK()) {
-          return retval;
-        }
-        continue;
-      }
-
-      auto expected_element_type = expected_type->AsTensorType()->GetElementType();
-      auto input_element_type = input_ml_value.Get<Tensor>().DataType();
-      auto retval = CheckTypes(input_element_type, expected_element_type);
-      if (!retval.IsOK()) {
-        return retval;
+      if (input_ml_value.IsTensor()) {
+        auto expected_element_type = expected_type->AsTensorType()->GetElementType();
+        auto input_element_type = input_ml_value.Get<Tensor>().DataType();
+        ORT_RETURN_IF_ERROR(CheckTypes(input_element_type, expected_element_type));
+      } else {
+        auto input_type = input_ml_value.Type();
+        ORT_RETURN_IF_ERROR(CheckTypes(input_type, expected_type));
       }
     }
-    return Status::OK();
-  }
 
-  common::Status ValidateInputNames(const std::vector<std::string>& feed_names) {
-    std::string missing_required_inputs;
+    if (feeds.size() > required_feed_ids.size()) {
+      // More feeds are offered.
+      // In the case of overriding some initializers (which are also taken as graph inputs).
+      for (size_t i = 0; i < feeds.size(); ++i) {
+        if (required_feed_ids.count(i) > 0) {
+          continue;
+        }
+        auto iter = input_def_map_.find(feed_names[i]);
+        if (input_def_map_.end() == iter) {
+          return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                                 "Invalid Feed Input Names: ", feed_names[i]);
+        }
 
-    const auto begin_names = feed_names.cbegin();
-    const auto end_names = feed_names.cend();
-    std::for_each(required_model_input_names_.cbegin(), required_model_input_names_.cend(),
-                  [&](const std::string& required_input) {
-                    if (std::find(begin_names, end_names, required_input) == end_names) {
-                      if (!missing_required_inputs.empty())
-                        missing_required_inputs += ",";
-
-                      missing_required_inputs += required_input;
-                    }
-                  });
-
-    if (!missing_required_inputs.empty()) {
-      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
-                             "Missing required inputs: ", missing_required_inputs);
-    }
+        auto& input_ml_value = feeds.at(i);
+        ORT_ENFORCE(input_ml_value.IsTensor());
```

Comment:
>ORT_ENFORCE [](start = 8, length = 11)

I'd suggest ORT_MAKE_STATUS for this. It's user input, and the user could provide an invalid MLValue. Our code is not broken, so I don't think ORT_ENFORCE is the best choice.